composite-checkout-wpcom: Checkout component specialized to WPCOM

nbloomf opened the pull request:
#### Changes proposed in this Pull Request

* Add cart management component for WPCOM

#### Testing instructions

Run tests: `npm install && npm run test-packages composite-checkout-wpcom`

Run live standalone demo: `npm run composite-checkout-wpcom-demo`

Enable in calypso: `ENABLE_FEATURES=composite-checkout-wpcom npm start`

Typecheck: `cd packages/composite-checkout-wpcom/ && tsc; cd ../..`

Commit:
Add composite-checkout-wpcom package

sirbrillig commented on packages/composite-checkout/src/components/wpcom/items.js:
```diff
@@ -0,0 +1,44 @@
+}
+
+export function DomainItem( { plan, item, onDeleteItem } ) {
+	const deleteItem = () => onDeleteItem( plan );
```
I think you meant to pass `item`  to `onDeleteItem()` since `plan` is not passed as a prop to this component.

sirbrillig commented on packages/composite-checkout/src/components/wpcom/items.js:
```diff
@@ -0,0 +1,44 @@
+					{ plan.subLabel && <div>{ plan.subLabel }</div> }
+				</span>
+				<span>{ renderDisplayValueMarkdown( plan.amount.displayValue ) }</span>
+				<button onClick={ deleteItem } />
```
If you add a string child to the `button` element, it will appear as the button text.

sirbrillig commented on packages/composite-checkout/src/components/wpcom/items.js:
```diff
@@ -0,0 +1,44 @@
+					{ item.subLabel && <div>{ item.subLabel }</div> }
+				</span>
+				<span>{ renderDisplayValueMarkdown( item.amount.displayValue ) }</span>
+				<button onClick={ deleteItem } />
```
Same comment about the missing button text.

sirbrillig commented on packages/composite-checkout/src/components/wpcom/edit-cart.js:
```diff
@@ -0,0 +1,6 @@
+export function adjustItemPricesForCountry() {}
+
+// replace this once https://github.com/Automattic/wp-calypso/pull/36758 is merged
+export function formatValueForCurrency() {}
```
Might as well replace this temporarily with something like:

```js
export function formatValueForCurrency( currency, value ) {
	const floatVersion = value / 100;
	return `${ floatVersion } ${ currency }`;
}
```

sirbrillig reviewed (commented):
Nice work so far!

sirbrillig commented on packages/composite-checkout/src/components/wpcom/checkout.js:
```diff
@@ -0,0 +1,68 @@
+			successRedirectUrl={ successRedirectUrl }
+			failureRedirectUrl={ failureRedirectUrl }
+			ReviewContent={ ReviewContent }
+			reviewContentCollapsed={ reviewContentCollapsed }
```
This prop no longer does anything. I forgot to remove it from the example.

sirbrillig commented on packages/composite-checkout/src/components/wpcom/checkout.js:
```diff
@@ -0,0 +1,68 @@
+			failureRedirectUrl={ failureRedirectUrl }
+			ReviewContent={ ReviewContent }
+			reviewContentCollapsed={ reviewContentCollapsed }
+			upSell={ upSell }
```
Let's remove `upSell` for now. Its API has changed and it's not really implemented yet anyway.

sirbrillig commented on packages/composite-checkout/src/components/wpcom/checkout.js:
```diff
@@ -0,0 +1,68 @@
+			locale={ 'US' }
+			items={ itemsWithTax }
+			total={ total }
+			onChangeBillingContact={ updatePricesForAddress }
```
Let's remove this for now because it's likely to work differently with the new data model and because we don't have the address update the price yet anyway.

sirbrillig commented on packages/composite-checkout/src/components/wpcom/checkout.js:
```diff
@@ -0,0 +1,68 @@
+			items={ itemsWithTax }
+			total={ total }
+			onChangeBillingContact={ updatePricesForAddress }
+			availablePaymentMethods={ availablePaymentMethods }
```
We can remove this and just allow all payment methods at the moment.

sirbrillig commented on packages/composite-checkout/src/components/wpcom/items.js:
```diff
@@ -0,0 +1,44 @@
+export function DomainItem( { plan, item, onDeleteItem } ) {
+	const deleteItem = () => onDeleteItem( plan );
+	return (
+		<React.Fragment>
```
This `Fragment` is not needed since we have a `div` here anyway.

sirbrillig reviewed (commented):
Looking good so far. Did you want to merge this work as-is, or try to add some `cart` transformation first?

sirbrillig commented on packages/composite-checkout/src/components/wpcom/translate-cart.js:
```diff
@@ -0,0 +1,162 @@
+		product_slug,
+		currency,
+		item_subtotal_integer, // TODO: implement this on the backend
+		item_subtotal_display, // TODO: implement this on the backend
```
For reference, D34860-code should add these.

sirbrillig commented on packages/composite-checkout/src/components/wpcom/checkout.js:
```diff
@@ -0,0 +1,63 @@
+			onFailure={ onFailure }
+			successRedirectUrl={ successRedirectUrl }
+			failureRedirectUrl={ failureRedirectUrl }
+			eventHandler={ handleCheckoutEvent }
```
You can remove this, btw. The new way to do this is to use `createRegistry()`, pass the result to `CheckoutProvider` as a the `registry` prop, and then use `registry.subscribe()` to subscribe to all changes to the store(s) (see the main demo for an example). That can probably wait, though, until we need to do something with those events.

sirbrillig commented:
I just merged #37501 without the feature flag enabled; I suggest that we modify this to replace the host page that was added in that PR.

sirbrillig commented on client/my-sites/checkout/checkout/composite-checkout-wpcom-container.js:
```diff
@@ -0,0 +1,75 @@
+
+const useShoppingCart = makeShoppingCartHook( callWpcomShoppingCartEndpoint( wpcom ), initialCart );
+
+export default function CompositeCheckoutWPCOMContainer() {
```
Let's just move this all over to `composite-checkout-container.js`, replace what's there, and rename this component `CompositeCheckoutContainer`.

sirbrillig commented on config/development.json:
```diff
@@ -38,6 +38,7 @@
 		"comments/filters-in-posts": true,
 		"comments/moderation-tools-in-posts": true,
 		"comments/management/threaded-view": false,
+		"composite-checkout-wpcom": true,
```
Let's remove this for now and force the user to set the flag manually.

sirbrillig commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -52,8 +52,8 @@ export function checkout( context, next ) {
 
-	if ( config.isEnabled( 'composite-checkout' ) ) {
-		context.primary = <CompositeCheckoutContainer />;
+	if ( config.isEnabled( 'composite-checkout-wpcom' ) ) {
```
Let's just remove all the changes in this file and rename your work to `CompositeCheckoutContainer`.

sirbrillig commented on packages/composite-checkout-wpcom/src/components/checkout.jsx:
```diff
@@ -0,0 +1,80 @@
+
+	return (
+		<CheckoutProvider
+			locale={ 'US' }
```
Let's change this to `en` for now which is more accurate even though we don't have any translations yet.

sirbrillig commented on packages/composite-checkout-wpcom/src/lib/edit-cart.js:
```diff
@@ -0,0 +1,8 @@
+export function adjustItemPricesForCountry() {}
+
+// replace this once https://github.com/Automattic/wp-calypso/pull/36758 is merged
+export function formatValueForCurrency( currency, value ) {
```
Let's add a safeguard here so we don't forget about this by adding `if ( currency !== 'USD' ) { throw new Error( 'Non-USD currency is not yet supported' ); }`

sirreal committed:
Use tsc to prepare package

sirreal committed:
Extend cjs from esm

sirreal committed:
Use esnext module for esm

sirbrillig commented on packages/composite-checkout-wpcom/src/components/checkout.jsx:
```diff
@@ -23,13 +23,11 @@ const failureRedirectUrl = window.location.href;
-	};
-}
+export const callWpcomShoppingCartEndpointWith = wpcom => async ( cartKey, cartParam ) => {
+	return await wpcom.setCart( cartKey, cartParam );
```
This function is `async`, which means it will always return a Promise. Inside it we are calling `wpcom.setCart()`, which also returns a Promise. When it's called, we're using `await` on that promise, which will wait for that Promise to complete before continuing. Once that Promise resolves, the outer function will return the resulting value in its own Promise, but that Promise will be resolved immediately.

If this were written without async/await, it would look like this:

```js
export const callWpcomShoppingCartEndpointWith = wpcom => ( cartKey, cartParam ) => {
  return new Promise( ( resolve, reject ) => {
    wpcom.setCart( cartKey, cartParam )
      .then( result => resolve( result ) )
      .catch( error => reject( error ) );
  } );
};
```

It's unnecessary to wait for the inner Promise to resolve before returning the outer Promise since there's no other execution that needs to occur after the inner Promise resolves. Instead of creating a second Promise, we could just return the inner Promise, like so:

```js
export const callWpcomShoppingCartEndpointWith = wpcom => ( cartKey, cartParam ) => {
  return wpcom.setCart( cartKey, cartParam );
};
```

Or, even as a one-liner:

```js
export const callWpcomShoppingCartEndpointWith = wpcom => ( cartKey, cartParam ) => 
  wpcom.setCart( cartKey, cartParam );
```

If you wanted to, you could even remove the forwarding function and just return the inner function since the arguments are the same:

```js
export const callWpcomShoppingCartEndpointWith = wpcom => wpcom.setCart;
```

nbloomf replied:
Thanks for the thorough explanation- that helps a lot. I was kind of flailing around putting async/await in different places trying to figure out where the server response was going, but it looks like your fix in ea9f8f5 addresses the real problem.

nbloomf replied:
Also- I like this a lot:

```
export const callWpcomShoppingCartEndpointWith = wpcom => wpcom.setCart;
```

However I'd also really like for the signature of this function to be manifest without needing to chase down where `wpcom.setCart` comes from. But while thinking about how to do that I realized that this function is only used in one place, and there's no need to pass the entire `wpcom` object in to this code- the whole point of injecting `wpcom.setCart` is to abstract out the call, and the `wpcom` object comes with... a lot of other stuff too. So I instead inlined it at the only call site like this:

```
const useShoppingCart = makeShoppingCartHook(
    ( cartKey, cartParam ) => wpcom.setCart( cartKey, cartParam ),
    initialCart
);
```

(It's probably possible to write that without the explicit arguments, but personally I prefer it that way in the absence of a static type signature)

sirbrillig replied:
I think you're absolutely right. It's a paint to have to search out the function definition to find out what the arguments are. 👍

sirbrillig commented on packages/composite-checkout/src/wpcom/wp-checkout-order-summary.js:
```diff
@@ -23,8 +23,8 @@ export default function WPCheckoutOrderSummary() {
 				<ProductList>
-					{ items.map( product => {
-						return <ProductListItem key={ product.id }>{ product.label }</ProductListItem>;
+					{ items.map( ( product, index ) => {
```
What is the `index` being used for here? The React docs [recommend against using it as a key](https://reactjs.org/docs/lists-and-keys.html#keys).

sirbrillig commented on packages/composite-checkout/src/wpcom/wp-checkout-order-summary.js:
```diff
@@ -23,8 +23,8 @@ export default function WPCheckoutOrderSummary() {
-					{ items.map( product => {
-						return <ProductListItem key={ product.id }>{ product.label }</ProductListItem>;
+					{ items.map( ( product, index ) => {
+						return <ProductListItem key={ product.label }>{ product.label }</ProductListItem>;
```
Are you sure that two products won't have the same label? The [schema for line items](https://github.com/Automattic/wp-calypso/blob/master/packages/composite-checkout/README.md#checkoutprovider) requires that each item have a unique `id` property, so using that might be best.

sirbrillig committed:
Undocumented: Allow getCart/setCart to return values

Each of these endpoint functions uses a wrapper (`_sendRequest`) to pass
their requests off to the server call made by `wpcom.req`. That latter
function has two modes: callback and Promise. The modern way to use the
request is in Promise mode, and is how most of the calls that use this
module work. However, these two functions do not return that Promise,
which means they only work in callback mode.

This commit changes both of those functions to return the Promise, if it
exists.

sirbrillig commented:
Nice work! There was a missing feature in `wpcom.undocumented.setCart`. I fixed it in ea9f8f5c94a48026ac4e8c40ab64d913b93a2ba2 and that will allow the server response to get back to you (see that commit for a more detailed explanation).
